### ISSUE TITLE
nemo_asr grammar seems to be incorrect

### ISSUE THREAD

**Issue by orrburgel** [NVIDIA/NeMo#3606]
**Describe the bug**

NeMo ASR grammar seems to be incorrect on any version except v1.4.0.

Besides the odd behavior with the grammar we can also notice that Punctuation takes much longer (100x~) in v1.4.0 compared to other versions that worked for my setup which leaves me with the only choice of using v1.4.0 for transcriptions and v1.6.0 for punctuation, eg:
```
NeMo Toolkit v1.4.0
Punctuation took: 13.06 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very like the old portrait."]
```
```
NeMo Toolkit v1.6.0
Punctuation took: 0.08 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
```

**Steps/Code to reproduce bug**

`pip install Cython`
```
pip install nemo_toolkit[all]==1.6.1
pip install nemo_toolkit[all]==1.6.0
pip install nemo_toolkit[all]==1.5.1
pip install nemo_toolkit[all]==1.4.0
```
I simply ran one of these every time I wanted to switch a version

**Expected behavior**

Let's inspect the debug log that I repeatedly tested multiple times on different versions of NeMo on the same wave file:
https://dldata-public.s3.us-east-2.amazonaws.com/2086-149220-0033.wav
If you will notice, only version 1.4.0 will get the word `like` right while all the others will sort of "make up" the word `likt` that doesn't really exists in any known dictionary, besides names, which I get the complexity of the manner, but am curios what "changed" or more like what got messed up so we can now see this behavior:

**NeMo Toolkit v1.4.0**

```
NeMo Toolkit v1.4.0
Init took: 27.58 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.14it/s]
Transcriptions took: 0.93 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very like the old portrait"]
Punctuation took: 13.77 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very like the old portrait."]
Everything took: 42.28 seconds

NeMo Toolkit v1.4.0
Init took: 29.06 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.15s/it]
Transcriptions took: 1.22 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very like the old portrait"]
Punctuation took: 12.90 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very like the old portrait."]
Everything took: 43.18 seconds

NeMo Toolkit v1.4.0
Init took: 28.26 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.29it/s]
Transcriptions took: 0.83 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very like the old portrait"]
Punctuation took: 12.20 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very like the old portrait."]
Everything took: 41.30 seconds

NeMo Toolkit v1.4.0
Init took: 27.88 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.16it/s]
Transcriptions took: 0.92 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very like the old portrait"]
Punctuation took: 13.06 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very like the old portrait."]
Everything took: 41.86 seconds
```

**NeMo Toolkit v1.5.1**

```
NeMo Toolkit v1.5.1
Init took: 24.72 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.00s/it]
Transcriptions took: 1.09 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 11.76batch/s]
Punctuation took: 0.10 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 25.91 seconds

NeMo Toolkit v1.5.1
Init took: 24.43 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.08it/s]
Transcriptions took: 1.01 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 11.36batch/s]
Punctuation took: 0.10 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 25.54 seconds

NeMo Toolkit v1.5.1
Init took: 24.63 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.06it/s]
Transcriptions took: 1.03 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 12.99batch/s]
Punctuation took: 0.09 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 25.76 seconds

NeMo Toolkit v1.5.1
Init took: 23.22 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.10it/s]
Transcriptions took: 0.99 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 14.29batch/s]
Punctuation took: 0.08 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 24.30 seconds
```

**NeMo Toolkit v1.6.0**

```
NeMo Toolkit v1.6.0
Init took: 22.43 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.27it/s]
Transcriptions took: 0.87 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 15.38batch/s]
Punctuation took: 0.07 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 23.37 seconds

NeMo Toolkit v1.6.0
Init took: 22.57 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.25it/s]
Transcriptions took: 0.88 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 15.38batch/s]
Punctuation took: 0.07 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 23.53 seconds

NeMo Toolkit v1.6.0
Init took: 22.14 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.22it/s]
Transcriptions took: 0.91 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 15.63batch/s]
Punctuation took: 0.07 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 23.12 seconds

NeMo Toolkit v1.6.0
Init took: 22.27 seconds
Transcribing: 100%|█████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.26it/s]
Transcriptions took: 0.88 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 15.15batch/s]
Punctuation took: 0.08 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 23.22 seconds
```

**NeMo Toolkit v1.6.1**

```
NeMo Toolkit v1.6.1
Init took: 37.51 seconds
Transcribing: 100%|██████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.30s/it]
Transcriptions took: 1.36 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|█████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.54s/batch]
Punctuation took: 1.58 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 40.45 seconds
```

**Environment overview**

 - Environment location: Local
 - Method of NeMo install: pip install. as can be seen above.

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS version: Latest
- PyTorch version: as required by each nemo package that I installed on a fresh machine using the commands above ^^
- Python version: 3.7.11 (also tested and got the same behavior on: 3.8.12 and 3.9.7)
- Python versions that didn't run it were: 2.7.18, 3.5.6, 3.6.13, 3.10.0

**Additional context**

I installed apex for cpu using:
```
pip install -v --disable-pip-version-check --no-cache-dir git+https://github.com/NVIDIA/apex
```
(TORCH_CUDA_ARCH_LIST="compute capability")

The following NeMo versions were not able to run at all using the same spec:
```
v1.3.0 and below (tested until v1.0.0)
v1.5.0
```

**Example Code:** (Fixed + added full snippet below)

```
import time
import nemo.collections.asr as nemo_asr
import nemo.collections.nlp as nemo_nlp

if __name__ == "__main__":
    print("NeMo Toolkit vX.Y.Z")
    start = time.time()
    
    asr_model = nemo_asr.models.EncDecCTCModel.from_pretrained("stt_en_quartznet15x5")
    punctuation = nemo_nlp.models.PunctuationCapitalizationModel.from_pretrained(model_name='punctuation_en_distilbert')
    print("Init took: " + '%.2f' % (time.time() - start) + " seconds")
    start = time.time()
    
    transcriptions = asr_model.transcribe(['2086-149220-0033.wav'])
    print("Transcription took: " + '%.2f' % (time.time() - start) + " seconds")
    start = time.time()
    
    res = punctuation.add_punctuation_capitalization(queries=transcriptions)
    print("Punctuation took: " + '%.2f' % (time.time() - start) + " seconds")
    
    print(transcriptions)
    print(res)
```

**Comment by titu1994** [comment]
Thank you for the exceptionally detailed issue ! Really helps us debug things. One question left out was what ASR model was used for these transcriptions? Even the code doesn't have a reference to which Nemo model was used for both asr and punctuation. 

We sometimes update checkpoints with the same reference I'd "stt_en_citrinet_1024_gamma_0_25" but train on mode data in the next version. So that can cause issues to show up. On the other hand, you should probably use newer models like Gamma Citrinets or Conformers. I don't believe we updated the punctuation model at all so I believe the issue is the update to the ASR model.

**Comment by titu1994** [comment]
Init took: 37.51 seconds
Transcribing: 100%|██████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.30s/it]
Transcriptions took: 1.36 seconds
["well i don't wish to see it any more observed phoebe turning away her eyes it is certainly very likt the old portrait"]
100%|█████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.54s/batch]
Punctuation took: 1.58 seconds
["Well, I don't wish to see it any more, observed Phoebe, turning away her eyes. It is certainly very likt the old portrait."]
Everything took: 40.45 seconds

Your init time is also 35+ seconds (!?), but the models are fast as every - 1.36 s for asr and 1.5 for punctuation. What took so long to initialize ?

**Comment by orrburgel** [comment]
@titu1994 The last test might be biased since I also had discord and a valheim server and also chrome running in the background :)
I'm glad I could help debug things thats my goal :)
Thank you for this awesome library I will soon read and respond with juicier details thanks :)

**Comment by orrburgel** [comment]
On first look I might have a silly question, how do I select the ASR model in my code? usually I make a list of all available models but couldn't find any for this yet

**Comment by orrburgel** [comment]
@titu1994 I found the missing lines my bad sorry:
```
    asr_model = nemo_asr.models.EncDecCTCModel.from_pretrained("stt_en_quartznet15x5")
    punctuation = nemo_nlp.models.PunctuationCapitalizationModel.from_pretrained(model_name='punctuation_en_distilbert')
```

**Comment by orrburgel** [comment]
@titu1994 You mentioned many models I'd like to explore please, like `stt_en_citrinet_1024_gamma_0_25` and etc, notice I use `stt_en_quartznet15x5` which makes me very curios is it as simple as swapping the name? if so may I have a link to a list of all of them `stt_en` models please?
I'm currently light years behind the main stream and would appreciate any guidance <3

**Comment by orrburgel** [comment]
Also I found genius ways to mute nltk_data warnings and the entire log besides the transcribe progress bar I would like to hide too, here is how I do it:
```
import logging
logging.disable()

import nltk
nltk_download = nltk.download
def _nltk_download(package, quiet=True):
    try:
        nltk.data.find('tokenizers/' + package)
    except LookupError:
        nltk_download(package, quiet=quiet)
nltk.download = _nltk_download

if __name__ == "__main__":
    print("NeMo Toolkit v1.6.1")

    import time
    gstart = time.time()
    start = time.time()
    def took(name, g=False):
        global start, gstart
        if g:
            print(name + " took: " + '%.2f' % (time.time() - gstart) + " seconds")
            gstart = time.time()
        else:
            print(name + " took: " + '%.2f' % (time.time() - start) + " seconds")
            start = time.time()

    import nemo.collections.asr as nemo_asr
    import nemo.collections.nlp as nemo_nlp

    asr_model = nemo_asr.models.EncDecCTCModel.from_pretrained("stt_en_quartznet15x5")
    punctuation = nemo_nlp.models.PunctuationCapitalizationModel.from_pretrained(model_name='punctuation_en_distilbert')

    took('Init')

    transcriptions = asr_model.transcribe(['2086-149220-0033.wav'])
    took('Transcriptions')
    print(transcriptions)

    res = punctuation.add_punctuation_capitalization(queries=transcriptions)
    took('Punctuation')
    print(res)

    took('Everything', True)
```
I hope you like it and that it help advance us

**Comment by orrburgel** [comment]
Why not, let's throw how I install NeMo on windows too:

```
conda install pywin32
pip install Cython

if torch.cuda.is_available():
pip install -v --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" --global-option="--fast_layer_norm" git+https://github.com/NVIDIA/apex
else:
set TORCH_CUDA_ARCH_LIST="compute capability"
pip install -v --disable-pip-version-check --no-cache-dir git+https://github.com/NVIDIA/apex

pip install nemo_toolkit[all]==1.6.1
pip install nemo_toolkit[all]==1.6.0
pip install nemo_toolkit[all]==1.5.1
pip install nemo_toolkit[all]==1.5.0
pip install nemo_toolkit[all]==1.4.0
pip install nemo_toolkit[all]==1.3.0

python testnemo.py

Cache: (301MB)
C:\Users\%username%\.cache\torch\NeMo\NeMo_V.E.R
```

**Comment by titu1994** [comment]
Nemo has always supported CPU - setting trainer.gpus = 0 and trainer.accelerator=None ensures that model will load and train and eval on CPU only. Doesnt matter what OS you are on. Windows is a surprise, but oh well.

>  asr_model = nemo_asr.models.EncDecCTCModel.from_pretrained("stt_en_quartznet15x5")

QuartzNet model is ancient.. As in its 3 years old (which in literature is classified as ancient). Conformers and Citrinets are much much better than these two for accuracy (and speed).

> makes me very curios is it as simple as swapping the name?

Yeah, just swap the name out and you get a new model. Theres documentation about this too, which you should read anyways cause it answers most questions - https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/results.html#ngc-pretrained-checkpoints
Theres also a page with all of this sorted by language - https://catalog.ngc.nvidia.com/orgs/nvidia/collections/nemo_asr

Multiple models, multiple languages, etc. You can use `"stt_<2 character language id>"` as substring match to select just the English models for example. We don't support all the languages of the world (yet), but the ones we do support are generally very good public checkpoints available. 

For even easier use do this
```
import nemo.collections.asr as nemo_asr

# This is for every model type, every language. You should subset these by language id.
all_models = nemo_asr.models.ASRModel. list_available_models()
for model_info in all_models:
    print(model_info)
```

------

Still - if you discount the startup time - all the models perform as expected on CPU 1-2 seconds for a single sample. So the only difference between all the versions is some have startup time and some don't. I don't see how nemo has any startup time as even on colab - it has install time yes but minute startup time (5-6 sec) on average to initialize cuda and parts of the library.
Sadly, I don't think we updated punctuation models to be more up to date.

**Comment by titu1994** [comment]
OOH wait, your init check - its accounting for NGC download time for the actual model. Ofc that is a network operation and depends on the speed of your internet etc. You should not count that as cost for running the model itself.

**Comment by orrburgel** [comment]
Thank you, I don't count Init as cost

**Comment by titu1994** [comment]
Each version of nemo will have a new cache of its model repo - so first time download cost is there. But after that there should be use from cache so it wont download again. You can rerun the experiment using cache to see actual runtime cost.

**Comment by orrburgel** [comment]
I was mainly wondering about the detection of the word 'likt' and how to get it to use the dictionary

**Comment by titu1994** [comment]
So ASR models will often either - 
1) Add characters they should not
2) Miss characters they should have put
3) Not print correct grammar because they don't have language model in them
4) not be able to detect text because of acoustic issues.

Many reasons why grammar can fail. No model is perfect. Not even google etc can get every word correct, that's why metrics like Word error rate and Character error rate exist.

**Comment by orrburgel** [comment]
Thank you so much again for this explanations you made me smarter @titu1994
